scipy/sparse/linalg/_special_sparse_arrays.py

```diff
@@ -0,0 +1,392 @@
+    >>> lap.sparse
+    <6x6 sparse matrix of type '<class 'numpy.int32'>'
+        with 20 stored elements in Compressed Sparse Row format>
+    >>> lap.toarray()
+    array([[-2.,  1.,  0.,  1.,  0.,  0.],
+           [ 1., -3.,  1.,  0.,  1.,  0.],
+           [ 0.,  1., -2.,  0.,  0.,  1.],
+           [ 1.,  0.,  0., -2.,  1.,  0.],
+           [ 0.,  1.,  0.,  1., -3.,  1.],
+           [ 0.,  0.,  1.,  0.,  1., -2.]])
+    >>> np.array_equal(lap(np.eye(n)), lap.toarray())
+    True
+    >>> np.array_equal(lap.tosparse().toarray(), lap.toarray())
+    True
+    >>> lap.eigenvalues
+    array([-5., -3., -3., -2., -1.,  0.])
+    >>> eigvals = np.sort(eigh(lap.toarray(), eigvals_only=True))
+    >>> np.allclose(lap.eigenvalues, eigvals)
+    True
+
+    """
+
+    def __init__(self, shape, *, boundary_conditions, dtype=np.int32) -> None:
+        self.dims = shape
+        self.bc = boundary_conditions
+        self.dtype = dtype
+        n = np.prod(shape)
+        super().__init__(dtype, (n, n))
+
+        indices = np.indices(shape)
+        L = np.zeros(shape)
+
+        for j, n in zip(indices, shape):
+            if boundary_conditions == "Dirichlet":
+                L += -4 * np.sin(np.pi * (j + 1) / (2 * (n + 1))) ** 2
+            elif boundary_conditions == "Neumann":
+                L += -4 * np.sin(np.pi * j / (2 * n)) ** 2
+            elif boundary_conditions == "Periodic":
+                L += -4 * np.sin(np.pi * np.floor((j + 1) / 2) / n) ** 2
+            else:
+                raise ValueError(
+                    f"Unknown value for {boundary_conditions}. The valid "
+                    "options are 'Dirichlet', 'Periodic', and 'Neumann'."
+                )
+        self.eigenvalues = np.sort(L.ravel())
+
+    def toarray(self):
+        dims = self.dims
+        bc = self.bc
+        N = np.prod(dims)
+        L = np.zeros([N, N], dtype=self.dtype)
+        # Scratch arrays
+        L_i = np.empty_like(L)
+        Ltemp = np.empty_like(L)
+
+        for ind, dim in enumerate(dims):
+            # Start zeroing out L_i
+            L_i[:] = 0
+            # Allocate the top left corner with the kernel of L_i
+            # Einsum returns writable view of arrays
+            np.einsum("ii->i", L_i[:dim, :dim])[:] = -2
+            np.einsum("ii->i", L_i[: dim - 1, 1:dim])[:] = 1
+            np.einsum("ii->i", L_i[1:dim, : dim - 1])[:] = 1
+
+            if bc == "Neumann":
+                L_i[0, 0] = -1
+                L_i[dim - 1, dim - 1] = -1
+            elif bc in ("Periodic", "Dirichlet"):
+                pass
+            else:
+                raise ValueError("Unknown boundary condition type "
+                                 f"'{bc}'.")
+
+            if bc == "Periodic":
+                if dim > 1:
+                    L_i[0, dim - 1] += 1
+                    L_i[dim - 1, 0] += 1
+                else:
+                    L_i[0, 0] += 1
```

Comment:
Also unify branch-handling for bc

```suggestion
            if self._bc == "Neumann":
                L_i[0, 0] = -1
                L_i[dim - 1, dim - 1] = -1
            elif self._bc == "Periodic":
                if dim > 1:
                    L_i[0, dim - 1] += 1
                    L_i[dim - 1, 0] += 1
                else:
                    L_i[0, 0] += 1
            # nothing to do for Dirichlet
```